### ISSUE TITLE
Allow users to rate pictures in their area

### ISSUE THREAD

**Issue by nicolas-raoul** [commons-app/apps-android-commons#2068]
From http://www.opensym.org/wp-content/uploads/2018/07/OpenSym2018_paper_30-1.pdf :

> “Photo Battle,” where instead of (or in
addition to) taking photographs, players could view and rate
the photos of other players in the same area. These ratings
could potentially then be aggregated and presented on the
Talk pages for the corresponding article, providing both
additional resources as well as augmenting commentary for
the original authors of those articles.

**Comment by Divyanshu2132** [comment]
Rating pictures is an good idea by this we can also remove images which are not of any use or if someone has uploaded his or her photo

**Comment by vanshikaarora** [comment]
Currently, do we have such an API?

**Comment by nicolas-raoul** [comment]
Structured Commons will introduce a "Quality assessment" property for each media.
Until it is ready, I suggest implementing the UI already, and as the backend just adding `==Android Commons app quality assessment== Score:4 (1:Low quality, 5: High quality)` to the media's talk page if the user selects 4 stars (for instance).

Also, let's add a "thank" to the picture if the number of stars is 4 or 5.

**Comment by vanshikaarora** [comment]
>Also, let's add a "thank" to the picture if the number of stars is 4 or 5.

Is it the same as https://github.com/commons-app/apps-android-commons/issues/2386 ?

**Comment by nicolas-raoul** [comment]
@vanshikaarora Yes both issues are about Mediawiki's "thanking" feature. The present issue would _give_ a thank, whereas the issue you mention would _receive_ the thank :-)

**Comment by vanshikaarora** [comment]
>@vanshikaarora Yes both issues are about Mediawiki's "thanking" feature. The present issue would _give_ a thank, whereas the issue you mention would _receive_ the thank :-)

That would be a great idea so currently shall we include the rating widget to the images in explore?

**Comment by nicolas-raoul** [comment]
Good question. I have not though about the UI.

If we mix featured pictures with normal pictures, users might rate normal pictures very harshly. Which might not be a problem, we could just maintain the user's medium rating for non-featured pictures, and apply a multiplication.

Maybe this activity could be sort of grouped with the peer review activity? Peers in the sense of mobile app users, and peers in the sense of same area. A "Peer review" item in the left menu would be acceptable.

Have you found the API that gives pictures in a current area? You can find it by looking at the JavaScript client-side code of https://tools.wmflabs.org/wikishootme/

**Comment by vanshikaarora** [comment]
>Until it is ready, I suggest implementing the UI already, and as the backend just adding ==Android Commons app quality assessment== Score:4 (1:Low quality, 5: High quality) to the media's talk page if the user selects 4 stars (for instance).

@nicolas-raoul where should the rating icon be added?

**Comment by vanshikaarora** [comment]
>we could just maintain the user's medium rating for non-featured pictures, and apply a multiplication.

**Comment by nicolas-raoul** [comment]
Ah yes, navigation drawer is the right word, thanks!

**Comment by vanshikaarora** [comment]
@nicolas-raoul Currently the "Peer-review" activity is will be showing a list of images along with their rating or is it something else?

**Comment by nicolas-raoul** [comment]
There are several ways to implement this.
A) In Explore, as a "Nearby pictures" tab. For consistency maybe the rating feature could be available on all pictures found in Explore, not just nearby pictures
B) As a part of Peer Review, for instance after recently uploaded pictures have all been reviewed (this would require creating a server-side database of what pictures have been reviewed).

**Comment by vanshikaarora** [comment]
>

**Comment by vanshikaarora** [comment]
>As a part of Peer Review, for instance after recently uploaded pictures have all been reviewed (this would require creating a server-side database of what pictures have been reviewed).

@nicolas-raoul Since currently we don't have the server-side ready. Shall I start with option A

**Comment by nicolas-raoul** [comment]
Yes if you want :-)
You can first implement the rating system on all pictures that are shown in Explore.

**Comment by nicolas-raoul** [comment]
For the UI, how about this:
1. When the user double-taps a picture, show this popup: ☆☆☆☆☆
2. When the user taps on the 4th star, it becomes: ★★★★☆
3. When the user taps OK, save the rating as explained at https://github.com/commons-app/apps-android-commons/issues/2068#issuecomment-461263882

**Comment by nicolas-raoul** [comment]
The same feature should also be available in the media details as double-tap too, and in the media details' menu (along with Download) for discoverability.

**Comment by vanshikaarora** [comment]
>Yes if you want :-)
You can first implement the rating system on all pictures that are shown in Explore.

Sure :)


>For the UI, how about this:
When the user double-taps a picture, show this popup: ☆☆☆☆☆
When the user taps on the 4th star, it becomes: ★★★★☆
When the user taps OK, save the rating as explained at #2068 (comment)

One point of suggestion here, user's might not know about this double tapping feature. So either,
 

- we can add a rate icon on the bottom/or on top of image clicking on which will display a popup.
- Or display a tutorial like this https://github.com/KeepSafe/TapTargetView only for the very first time. 


Please share your opinion  @nicolas-raoul  :)

**Comment by nicolas-raoul** [comment]
A rate icon near the download icon or inside the menu in the media details would be good, I think.
A tutorial would be too much intrusion for a non-essential feature, I think.

**Comment by vanshikaarora** [comment]
>A rate icon near the download icon or inside the menu in the media details would be good

Ok I'll add the rate icon in the media details :+1:

**Comment by vanshikaarora** [comment]
Task List for me:

- [ ] Adding UI as discussed [here](https://github.com/commons-app/apps-android-commons/issues/2068#issuecomment-473532644) 
- [ ] Adding Android Commons app quality assessment== Score:4 (1:Low quality, 5: High quality) to the media's talk page
- [ ] Add a "thank" to the picture if the number of stars is 4 or 5.

**Comment by vanshikaarora** [comment]
@nicolas-raoul As discussed [here](https://github.com/commons-app/apps-android-commons/issues/2425#issuecomment-472791645). Currently we do not have the an API to write media talk page.  Task 2 in my work list shall be kept on hold. Please correct me if I am wrong :)

>Add a "thank" to the picture if the number of stars is 4 or 5.

Is it  like sending a thanks to the editor of the picture?

**Comment by nicolas-raoul** [comment]
# Talk page
I think this is the API to use to edit any wiki article (which includes Commons talk pages):
https://github.com/wikimedia/wikimedia-android-data-client/blob/master/src/main/java/org/wikipedia/edit/EditClient.java#L25
You need to merge https://github.com/commons-app/apps-android-commons/pull/2642 into your branch to start using it.

# Thank
The app must thank the wiki edit which is the upload of that media. Just like clicking the "thank" link at https://commons.wikimedia.org/w/index.php?title=File:Baragazza_vista_del_centro.jpg&action=history :
![Screenshot from 2019-03-18 14-33-21](https://user-images.githubusercontent.com/99590/54508958-2a45a580-498b-11e9-9938-9a31eb05f77d.png)

**Comment by vanshikaarora** [comment]
>You need to merge #2642 into your branch to start using it.

Do we need to further editing on this PR? Or it is sufficient to be merged with my branch?

**Comment by nicolas-raoul** [comment]
That PR will get more commits, but even now it should be enough to edit talk pages I believe.

**Comment by vanshikaarora** [comment]
@nicolas-raoul So shall I merge the PR in it's current state with my branch?

**Comment by nicolas-raoul** [comment]
Yes, merge that PR into your branch.
Please first commit all of your changes first so that you can easily reset if things go wrong.

**Comment by misaochan** [comment]
Was this implemented?

**Comment by nicolas-raoul** [comment]
@misaochan Not really. Thanks were implemented, by not "in your area". Maybe it could be another tab in Explore, after "Mobile uploads"?

I suggest abandoning the idea of star rating for now, and instead making any picture targetable for "peer review", where the user can either thank or tag any problem.

**Comment by misaochan** [comment]
> "in your area". Maybe it could be another tab in Explore, after "Mobile uploads"?

This would be fantastic IMO. :)
